src/sentry/services/hybrid_cloud/auth/impl.py

```diff
@@ -76,8 +105,33 @@ def query_sso_state(
 
 
 class DatabaseBackedAuthService(AuthService):
-    # Monolith implementation that uses OrganizationMember and User tables freely, but in silo world
-    # this won't be possible.
+    def authenticate(self, *, request: AuthenticationRequest) -> AuthenticationResponse:
+        fake_request = FakeAuthenticationRequest(request)
+        handler: Any = RequestAuthenticationMiddleware()
+        expired_user: User | None = None
+        try:
+            handler.process_request(fake_request)
```

Comment:
What led to this approach? Shouldn't middleware being using the services? This looks like the service depending on middleware instead.

Reply:
In the end, the fact that `RequestAuthenticationMiddleware` is middleware, is just a convenience of its usage.  It's basically, *just a function that accepts and mutates a request object with a resulting user and auth value*.

So if it helps, I can drop the middleware part of this and make it a pure function instead.

Reply:
Ultimately, one way to think of this is, is that the control silo *is the definitive authority on authentication of a user*.  Ensuring that 100% of that logic is run inside of the service layer means that there is no risk of code drift from authenticating via an region silo.  Only the latest control silo providers authentication, and does so by providing the entirity of its encapsulated logic.

Service domains over service objects.

Reply:
>  Forcing that django logic to use the service means vendoring the entire django authentication stack (lots of functions, not just one or two).

Would what we need to do not fit within Django's [authentication backends](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#writing-an-authentication-backend)? 

We could evolve our implementation towards a Django authentication backend in the future.

> Performance. A single RPC call that can capture all of the logic of authentication in the control silo was preferable to many many tiny little RPC calls sprinkled around for various objects.

That totally makes sense. Cross silo authentication will need to be a single RPC call in the future we can't take the performance overhead of multiple round trips.

Reply:
> Would what we need to do not fit within Django's [authentication backends](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#writing-an-authentication-backend)?
We could evolve our implementation towards a Django authentication backend in the future.

We actually already do implement backends -- several of them.  Some of them return users, but some return social users.  some of them are intended as authentication and some of them are hammered in as suppport logic. Then we call back into django authentication helpers that assume a certain interface and implementation, which might include trying to fetch users on its own.  And then we do some more logic, like calling for Apitokens, etc, etc.   And the entry middelware, btw, has its own custom logic that won't fit into a backend altogether because of additional state and flow it carries. 

I, don't think we can re-use the django framework to solve this problem by the end.  We either put our service context outside  and above the spiderweb that is our authentication framework, and be forced to use a fake request object to perform the work, or we vendor and reimplement our own authentication framework.  Like, I've definitely tried otherwise and it didn't end well.

Reply:
Oh, I also forgot how our authentication system currently uses *lazy user objets* to fit into the circular dependency with `Superuser`.  :///